File: scripts/test_source_counts.mjs
#!/usr/bin/env node
// The layer list's feature counts belong to the STATE, not to app.js.
//
// A vector tile carries no global count, so the totals shown beside each
// overlay layer are baked at build time -- and they were baked into SOURCES in
// app.js as literals. Those literals were Washington's, so the moment a second
// state shipped, Oregon's layer list reported Washington's numbers: 55,271 BLTS
// segments against its actual 81,210, and 338,650 roads against 308,347.
//
// Nothing caught it because nothing compared a displayed number to the data
// behind it. This does, for every state, against the files that ship.
import { readFileSync } from 'node:fs';
import { createRequire } from 'node:module';
import zlib from 'node:zlib';
import { ROOT, check, checkEqual, done, source } from './testlib/harness.mjs';

const { MAP_STATES } = createRequire(import.meta.url)(ROOT + '/maps/states.js');
const appSrc = source('app.js');

// SOURCES must read Region, not carry literals. A number here is a state fact
// in shared code, which is the bug this file exists for.
const literals = appSrc.match(/^\s*count: \d+/gm) || [];
check('no layer count is hardcoded in app.js', literals.length === 0,
  literals.join(' | '));

const features = (path) => {
  const raw = zlib.gunzipSync(readFileSync(path));
  return JSON.parse(raw.toString('utf8')).features.length;
};

for (const state of MAP_STATES) {
  const counts = state.sourceCounts || {};
  check(`${state.id}: declares its layer counts`,
    !!counts.blts && !!counts.bikeinfra && !!counts.roads, JSON.stringify(counts));
  if (!counts.blts) continue;

  // blts ships whole, so this is exact.
  if (state.datasets?.overlays) {
    checkEqual(`${state.id}: the BLTS count matches blts.geojson.gz`,
      counts.blts, features(`${ROOT}/maps/${state.id}/blts.geojson.gz`));
  }
  // bikeinfra is filtered before tiling (sharrow-only ways are dropped), so the
  // shipped total is only an upper bound here. test_source_counts.py PINS it
  // exactly, by importing the real sharrow_only() from the builder rather than
  // reimplementing the rule in a second language -- which would be the very
  // defect this family of tests exists to catch.
  const infra = features(`${ROOT}/maps/${state.id}/bikeinfra.geojson.gz`);
  check(`${state.id}: the bike-infrastructure count is within its own data`,
    counts.bikeinfra > 0 && counts.bikeinfra <= infra,
    `${counts.bikeinfra} declared vs ${infra} shipped`);
}

// The counts must DIFFER between states. If a second state is ever given the
// first one's numbers again, this is the check that says so.
const seen = MAP_STATES.filter((s) => s.sourceCounts?.blts)
  .map((s) => `${s.sourceCounts.blts}/${s.sourceCounts.roads}`);
check('no two states report the same counts', new Set(seen).size === seen.length,
  seen.join(' | '));

done();

File: scripts/test_source_counts.py
#!/usr/bin/env python3
"""The bike-infrastructure count in region.json is the one the tiler produces.

test_source_counts.mjs pins the BLTS and roads totals, but it could only BOUND
this one: bikeinfra is filtered before tiling -- sharrow-only ways are dropped
-- and reproducing that filter in JavaScript would mean a second copy of the
rule, which is the exact defect this family of tests exists to catch.

So this lives in Python and imports the real sharrow_only() from the builder.
One rule, one place, and the number a rider reads in the layer list is checked
against the data that ships.

I claimed this needed a full build_overlay_tiles.py run over the pipeline. It
does not. The builder reads maps/<state>/bikeinfra.geojson.gz, which is
committed; only the TILING step needs tippecanoe, and the count does not.
"""
import gzip
import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import build_overlay_tiles  # noqa: E402

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
failures = 0


def check(name, ok, detail=''):
    global failures
    if ok:
        print(f'  ok   {name}')
    else:
        failures += 1
        print(f'  FAIL {name}' + (f'  -- {detail}' if detail else ''))


states = json.load(open(os.path.join(ROOT, 'maps', 'states.json'))) \
    if os.path.exists(os.path.join(ROOT, 'maps', 'states.json')) else None
if states is None:
    states = []
    maps = os.path.join(ROOT, 'maps')
    for name in sorted(os.listdir(maps)):
        config = os.path.join(maps, name, 'region.json')
        if os.path.isfile(config):
            states.append(json.load(open(config)))

for state in states:
    counts = state.get('sourceCounts') or {}
    path = os.path.join(ROOT, 'maps', state['id'], 'bikeinfra.geojson.gz')
    if not counts.get('bikeinfra') or not os.path.exists(path):
        continue
    with gzip.open(path) as fh:
        data = json.load(fh)
    kept = sum(1 for f in data['features']
               if not build_overlay_tiles.sharrow_only(f['properties']))
    check(f"{state['id']}: the bike-infrastructure count is what the tiler keeps",
          counts['bikeinfra'] == kept,
          f"region.json says {counts['bikeinfra']:,}, the filter keeps {kept:,}")

if not failures:
    print(f'\n{len(states)} states checked, 0 failed')
sys.exit(1 if failures else 0)
